did_finder_rucio/src/rucio_did_finder/replica_distance.py

```diff
@@ -0,0 +1,159 @@
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import logging
+import os
+
+from typing import List, Mapping, Optional, Tuple
+from socket import gethostbyname
+import math
+from functools import lru_cache
+import tempfile
+from urllib.parse import urlparse
+import geoip2.database
+import geoip2.errors
+
+logger = logging.getLogger('ReplicaDistanceService')
+
+
+def _haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float):
+    ''' Assume inputs are in degrees; will convert to radians. Returns distance in radians '''
+    dellat = math.radians(lat2-lat1)
+    dellon = math.radians(lon2-lon1)
+    hav_theta = ((1-math.cos(dellat))/2 +
+                 math.cos(math.radians(lat1))*math.cos(math.radians(lat2))*(1-math.cos(dellon))/2)
+
+    return 2*math.asin(math.sqrt(hav_theta))
+
+
+@lru_cache
+def _get_distance(reader: Optional[geoip2.database.Reader],
+                  fqdn: str, my_lat: float, my_lon: float):
+    """
+    Determine angular distance between server at fqdn and (my_lat, my_lon).
+    If there is a failure of fdqn location lookup, will return pi
+    """
+    if reader is None:
+        return math.pi
+    try:
+        loc_data = reader.city(gethostbyname(fqdn)).location
+    except geoip2.errors.AddressNotFoundError as e:
+        logger.warning(f'Cannot geolocate {fqdn}, returning maximum distance.\nError: {e}')
+        return math.pi
+    site_lat, site_lon = loc_data.latitude, loc_data.longitude
+    if site_lat is None or site_lon is None:
+        return math.pi
+    return _haversine_distance(site_lat, site_lon, my_lat, my_lon)
+
+
+class ReplicaSorter(object):
+    _reader: Optional[geoip2.database.Reader] = None
+    # we keep the temporary directory around so it won't get randomly deleted by the GC
+    _tmpdir: Optional[tempfile.TemporaryDirectory] = None
+
+    def __init__(self, db_url_tuple: Optional[Tuple[str, bool]] = None):
+        """
+        Argument is an optional tuple of (URL, bool).
+        The URL is assumed to be a file to download; the bool indicates whether
+        it is ready to be used (True) or needs unpacking (False)
+        """
+        if db_url_tuple is None:
+            db_url_tuple = self.get_download_url_from_environment()
+        self._download_data(db_url_tuple)
+
+    def sort_replicas(self, replicas: List[str], location: Mapping[str, float]) -> List[str]:
+        """
+        Main method of this class.
+        replicas: list of strings which are the URLs for the replicas for a file
+        location: dict of the form {'latitude': xxx, 'longitude': yyy} where xxx and yyy are floats
```

Comment:
Any reason lat and long can't be instance variables?

Reply:
When originally written, the idea was you could have one ReplicaSorter which would be a service accepting distance sorting requests from clients (i.e. subsequent calls could have different locations), so this is a bit of an artifact. That said since this is now internal to the Rucio DID finder I don't think it's particularly interesting where exactly we store the numbers.